elfo-network/src/connman.rs

```diff
@@ -0,0 +1,346 @@
+                    },
+                    _ => id,
+                };
+
+                Some((st, id))
+            })
+            // Simplest way of draining whole iterator, it's needed for
+            // ID reallocation in case of accidental iterator drop.
+            .collect::<Vec<_>>();
+        (least_time, ready.into_iter())
+    }
+
+    pub(crate) fn insert_new(
+        &mut self,
+        role: ConnectionRole,
+        transport: ConnectTransport,
+    ) -> ManagedConn<'_> {
+        let id = self.conns.insert(Conn {
+            role,
+            transport,
+            state: State::New(NewState {
+                connect_at: Instant::now(),
+            }),
+        });
+        // SAFETY: just inserted.
+        unsafe { self.get_mut(id).unwrap_unchecked() }
+    }
+
+    pub(crate) fn insert_established(
+        &mut self,
+        role: ConnectionRole,
+        transport: ConnectTransport,
+    ) -> ManagedConn<'_> {
+        let id = self.conns.insert(Conn {
+            role,
+            transport,
+            state: State::Established,
+        });
+
+        self.get_mut(id).expect("just inserted")
+    }
+
+    pub(crate) fn get_mut(&mut self, id: ConnId) -> Result<ManagedConn<'_>, NoSuchConn> {
+        self.conns
+            .get_mut(id)
+            .map(|conn| ManagedConn {
+                id,
+                conn,
+                config: &self.config,
+            })
+            .ok_or(NoSuchConn(id))
+    }
+
+    /// Reallocate ID for that connection, previous id would be
+    /// no longer accessible.
+    pub(crate) fn realloc_id(&mut self, prev: ConnId) -> Result<ManagedConn<'_>, NoSuchConn> {
+        let prev = self.conns.remove(prev).ok_or(NoSuchConn(prev))?;
+        let id = self.conns.insert(prev);
+
+        Ok(self.get_mut(id).expect("just inserted"))
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct NewState {
+    connect_at: Instant,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct FailedState {
+    reconnect_at: Instant,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, IsVariant, derive_more::Display)]
+pub(crate) enum State {
+    #[display("New")]
+    New(NewState),
+    Establishing,
+    Established,
+    Accepted,
+    #[display("Failed")]
+    Failed(FailedState),
+}
+
+#[derive(Debug, derive_more::Display)]
+#[display("{state}({role}, {transport})")]
+pub(crate) struct Conn {
+    role: ConnectionRole,
+    state: State,
+    transport: ConnectTransport,
+}
+
+impl Conn {
+    pub(crate) const fn transport(&self) -> &ConnectTransport {
+        &self.transport
+    }
+
+    pub(crate) const fn role(&self) -> ConnectionRole {
+        self.role
+    }
+
+    pub(crate) const fn state(&self) -> State {
+        self.state
+    }
+
+    fn switch_state(&mut self, new: State) -> State {
+        let prev = std::mem::replace(&mut self.state, new);
+        debug!(
+            from = %prev,
```

Comment:
Still useless log. I prefer to see one log per state switch with info about peer, socketinfo, prev/new status